doc/manual/src/command-ref/cli-guideline.md

```diff
@@ -0,0 +1,571 @@
+# CLI guideline
+
+## Goals
+
+Purpose of this document is to provide a clear direction to **help design
+delightful command line** experience. This document contain guidelines to
+follow to ensure a consistent and approachable user experience.
+
+## Overview
+
+`nix` command provides a single entry to a number of sub-commands that help
+**developers and system administrators** in the life-cycle of a software
+project. We particularly need to pay special attention to help and assist new
+users of Nix.
+
+# Naming the `COMMANDS`
+
+Words matter. Naming is an important part of the usability. Users will be
+interacting with Nix on a regular basis so we should **name things for ease of
+understanding**.
+
+We recommend following the [Principle of Least
+Astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).
+This means that you should **never use acronyms or abbreviations** unless they
+are commonly used in other tools (e.g. `nix init`). And if the command name is
+too long (> 10-12 characters) then shortening it makes sense (e.g.
+“prioritization” → “priority”).
+
+Commands should **follow a noun-verb dialogue**. Although noun-verb formatting
+seems backwards from a speaking perspective (i.e. `nix store copy` vs. `nix
+copy store`) it allows us to organize commands the same way users think about
+completing an action (the group first, then the command).
+
+## Naming rules
+
+Rules are there to guide you by limiting your options. But not everything can
+fit the rules all the time. In those cases document the exceptions in [Appendix
+1: Commands naming exceptions](#appendix-1-commands-naming-exceptions) and
+provide reason. The rules want to force a Nix developer to look, not just at
+the command at hand, but also the command in a full context alongside other
+`nix` commands.
+
+```shell
+$ nix [<GROUP>] <COMMAND> [<ARGUMENTS>] [<OPTIONS>]
+```
+
+- `GROUP`, `COMMAND`, `ARGUMENTS` and `OPTIONS` should be lowercase and in a
+  singular form.
+- `GROUP` should be a  **NOUN**.
+- `COMMAND` should be a **VERB**.
+- `ARGUMENTS` and `OPTIONS` are discussed in [*Input* section](#input).
+
+## Classification
+
+Some commands are more important, some less. While we want all of our commands
+to be perfect we can only spend limited amount of time testing and improving
+them.
+
+This classification tries to separate commands in 3 categories in terms of
+their importance in regards to the new users. Users who are likely to be
+impacted the most by bad user experience. This does not mean that we will not
+take care of Tier3 commands, it means we will only put more helpful details for
+beginners into Tier1 commands.
+
+- **Tier1**: Commands used for our main use cases and most likely used by new
+  users. From Tier1 commands we expect attention to details, such as:
+    - Proper use of [colors](#colors), [emojis](#special-unicode-characters)
```

Comment:
I did not know there was a proper use of emoji already :slightly_smiling_face:

Reply:
Separately, while emojis are described for conveying emotions which is not the first thing I expect from a command-line tool, and also outside the range of many nice fonts in case of single-font configuration of a terminal emulator, there are pictograms that are in BMP, have better support, and a lot of precedent of use in technical writing!

Reply:
As with everything the balance is the key. I guess you were alerted by the word emojis  :) and seeing them all over your terminal. I'm also annoyed by command line tools that abuse this. But I have also experience command line tools that know how to use emojis.

Like it or not we are all emotional beings. Most of our decisions are based on emotions. And this is why things that we like to use everyday also convey certain emotions. Why wouldn't command line tools do the same.

Just to make sure you understand for what I am advocating I'm going to repeat it in the best ever (:wink:) format to explain things, single bullet points:

- We should use emojis in only few places: (1) this is because some find them annoying (2) with every emoji the rest of the emojis _"lose on power"_.  And when I say few places I really mean only two, maybe maybe tree places.

So key is balance and rather to not use emojis then to use emojis, but there might be few places where usage of emojis makes Nix better. Of course there is a way to disable this altogether and we should even disable it automatically (eg. when tty is not attached to stdout/stderr).